myst_nbs/gaussian_processes/gaussian_process.myst.md

```diff
@@ -135,45 +130,53 @@ with gp_fit:
     D_off_diag = squared_distance(x, xgrid)
 
     # Covariance matrices for prediction
-    Σ_pred = η_sq * T.exp(-ρ_sq * D_pred)
-    Σ_off_diag = η_sq * T.exp(-ρ_sq * D_off_diag)
+    sigma_pred = eta_sq * at.exp(-rho_sq * D_pred)
+    sigma_off_diag = eta_sq * at.exp(-rho_sq * D_off_diag)
 
     # Posterior mean
-    μ_post = Deterministic("μ_post", T.dot(T.dot(Σ_off_diag, matrix_inverse(Σ)), y))
+    mu_post = pm.Deterministic("mu_post", at.dot(at.dot(sigma_off_diag, matrix_inverse(sigma)), y))
     # Posterior covariance
-    Σ_post = Deterministic(
-        "Σ_post", Σ_pred - T.dot(T.dot(Σ_off_diag, matrix_inverse(Σ)), Σ_off_diag.T)
+    sigma_post = pm.Deterministic(
+        "sigma_post",
+        sigma_pred - at.dot(at.dot(sigma_off_diag, matrix_inverse(sigma)), sigma_off_diag.T),
     )
 ```
 
 ```{code-cell} ipython3
 with gp_fit:
-    svgd_approx = pm.fit(400, method="svgd", inf_kwargs=dict(n_particles=500))
+    svgd_approx = pm.fit(400, method="svgd", inf_kwargs=dict(n_particles=100))
 ```
 
 ```{code-cell} ipython3
-gp_trace = svgd_approx.sample(1000)
+gp_trace = svgd_approx.sample(1000).stack(sample=("chain", "draw"))
 ```
 
 ```{code-cell} ipython3
-traceplot(gp_trace, var_names=["η_sq", "ρ_sq", "σ_sq"]);
+az.plot_trace(gp_trace, var_names=["eta_sq", "rho_sq", "sigma_sq"]);
 ```
 
-Sample from the posterior GP
+Sample from the posterior Gaussian Process
 
 ```{code-cell} ipython3
 y_pred = [
-    np.random.multivariate_normal(m, S) for m, S in zip(gp_trace["μ_post"], gp_trace["Σ_post"])
+    np.random.multivariate_normal(m, S)
+    for m, S in zip(gp_trace.posterior["mu_post"].T, gp_trace.posterior["sigma_post"].T)
```

Comment:
This is very frustrating because neither numpy nor scipy support generating random draws from a multivariate normal in a batched way. I do know they are working on this though and was planning on waiting until this was fixed to add the multivariate distributions to https://einstats.python.arviz.org/en/latest/api/stats.html. But it wouldn't be very difficult to add a multivariate normal batched random number generator (using the already available standard normal and cholesky functions in einstats).

Is that something that would help? It would obviate the need for the stack a few lines above and for the loop.

Reply:
Extra note: I also think that given how our only goal is plotting, generating `y_pred` values for all posterior draws is a bit too much. The plot later on is a bit too crowded. I think that overlaying 100-200 draws is already enough. 

If you install the [ArviZ development version](https://python.arviz.org/en/latest/getting_started/Installation.html#development) you will be able to do something like:

```
post = az.extract(idata, num_samples=100, rng=rng)
# here loop or use einstats or something else, the extract function defaults to stacking under the hood too
```

Reply:
`stack` was removed and the for was removed and now `y_pred` is calculated with a vectorized function to remove the `for.`